nodejs/awsx/apigateway/authorizer.ts

```diff
@@ -0,0 +1,355 @@
+     * While this argument is optional, at least one queryParameter or one header must be defined.
+     * */
+    queryParameters?: string[];
+
+    /**
+     * headers is an array of the expected header keys used to authorize a request.
+     *  While this argument is optional, at least one queryParameter or one header must be defined.
+     * */
+    headers?: string[];
+
+    /**
+     * The authorizerHandler specifies information about the authorizing Lambda. You can either set
+     * up the Lambda separately and just provide the required information or you can define the
+     * Lambda inline using a JavaScript function.
+     */
+    handler: LambdaAuthorizerInfo | aws.lambda.EventHandler<AuthorizerEvent, AuthorizerResponse>;
+
+    /**
+     * The number of seconds during which the resulting IAM policy is cached. Default is 300s. You
+     * can set this value to 0 to disable caching. Max value is 3600s. Note - if you are sharing an
+     * authorizer across more than one route you will want to disable the cache or else it will
+     * cause problems for you.
+     */
+    authorizerResultTtlInSeconds?: number;
+}
+
+/**
+ * getRequestLambdaAuthorizerDefinition is a helper function to generate a request
+ * LambdaAuthorizerDefinition.
+ * @param name - the name for the authorizer. This must be unique for each unique authorizer in the
+ * API.
+ * @param args - configuration information for the token Lambda.
+ */
+export function getRequestLambdaAuthorizerDefinition(args: RequestAuthorizerArgs): LambdaAuthorizerDefinition {
+    let parameterName: string;
+    let location: "header" | "query";
+
+    const numQueryParams = getLength(args.queryParameters);
+    const numHeaders = getLength(args.headers);
+
+    if (numQueryParams === 0 && numHeaders === 0) {
+        throw new Error("Must specify at least one parameter");
+    } else {
+        location = getLocation(numHeaders, numQueryParams);
+        parameterName = getParameterName(args, numHeaders, numQueryParams);
+    }
+
+    return {
+        authorizerName: args.authorizerName,
+        parameterName: parameterName,
+        parameterLocation: location,
+        authType: "custom",
+        type: "request",
+        handler: args.handler,
+        identitySource: parametersToIdentitySources(args),
+        authorizerResultTtlInSeconds: args.authorizerResultTtlInSeconds,
+    };
+}
+
+/** @internal */
+function getLength(params: string[] | undefined): number {
+    if (!params) {
+        return 0;
+    }
+    return params.length;
+}
+
+/** @internal */
+function getParameterName(args: RequestAuthorizerArgs, numHeaders: number, numQueryParameters: number): string {
+    if (numQueryParameters + numHeaders === 1) {
+        if (args.queryParameters) {
+            return args.queryParameters[0];
+        } else if (args.headers) {
+            return args.headers[0];
+        }
+    }
+    return "Unused";
```

Comment:
intententional?  what are the semantics if both are provided?

Reply:
this is intentional -- if more than 1 is provided this must be "Unused"